jmh-core/src/main/java/org/openjdk/jmh/profile/LinuxPerfAsmProfiler.java

```diff
@@ -46,17 +46,28 @@ public class LinuxPerfAsmProfiler extends AbstractPerfAsmProfiler {
     public LinuxPerfAsmProfiler(String initLine) throws ProfilerException {
         super(initLine, "cycles");
 
-        String[] senseCmd = { PerfSupport.PERF_EXEC, "stat", "--event", Utils.join(requestedEventNames, ","), "--log-fd", "2", "echo", "1" };
+        String[] senseCmd = { PerfSupport.PERF_EXEC, "record", "--event", Utils.join(requestedEventNames, ","), "--output", "perf-record-validate.data", "echo", "1" };
 
         Collection<String> failMsg = Utils.tryWith(senseCmd);
         if (!failMsg.isEmpty()) {
             throw new ProfilerException(failMsg.toString());
         }
 
-        Collection<String> passMsg = Utils.runWith(senseCmd);
         for (String ev : requestedEventNames) {
-            if (PerfSupport.containsUnsupported(passMsg, ev)) {
-                throw new ProfilerException("Unsupported event: " + ev);
+            String reportCmd = String.format("perf report -i perf-record-validate.data --stdio | grep %s | wc -l", ev);
```

Comment:
> Why can't we just use perf record --event $requestedEventNames --output $tempdir echo 1? 

I couldn't find a reliable way that worked bare metal and on CI that would allow me to inspect the output of that and validate whether it worked or not. I've only been able to make it work in all environments by explicitly adding an output file (with `-i`) and then using `perf report` to verify its contents.

What are the issues I had? If you look at the commit series you can see the progression of trial/error.

First I tried to use `perf record` and see if I could just parse its output to see if could find a way to make sure it worked fine. I tried the approach [here](https://github.com/galderz/jmh/blob/2a9380664f9dd667fa10fe08b12d5c7e97c58b95/jmh-core/src/main/java/org/openjdk/jmh/profile/LinuxPerfAsmProfiler.java#L58), but that didn't work on [CI run](https://github.com/openjdk/jmh/runs/28853969690) for it. I'm not 100% sure why it didn't work, but I found odd that there was so much additional output in the `perf record` output, see my tests [here](https://github.com/galderz/github-actions/actions/runs/10417234664/job/28851057186).

After some test I realised that the additional output in `perf record` is the output that normally goes into `perf.data` file but on CI that does not happen. Instead the output goes to standard output.

So, if the output of `perf record` cannot really be used to determine the outcome, I decided to combine it with `perf report`. Also, I explicitly added the `-i` parameter to add an output file and avoid the mangling of the output.

Eventually I narrowed down the solution to the combination of `perf record` and `perf report` commands that you see in this [GH action](https://github.com/galderz/github-actions/actions/runs/10418351274/job/28854335728) of mine. That's the logic I applied to the code here and seems to work as expected.

Reply:
Also note what happens [here](https://github.com/galderz/github-actions/actions/runs/10418351274/job/28854335728).

The command executed is:

$ perf record --event cycles --output perf-record-validate.data echo 1
And the output contains:

[ perf record: Captured and wrote 0.001 MB perf-record-validate.data (3 samples) ]
So you can't use my logic [here](https://github.com/galderz/jmh/blob/2a9380664f9dd667fa10fe08b12d5c7e97c58b95/jmh-core/src/main/java/org/openjdk/jmh/profile/LinuxPerfAsmProfiler.java#L58) to decide whether perf record worked or not by checking if the output has [ perf record: Captured and wrote 0.000 MB (null) ]. So, how do you expect to sense that the perf record command worked from that alone?

Reply:
Also, you could have the situation that you pass in multiple events and a subset is only supported. To validate that you have to process the `perf record` output and see if there are events for all of them, otherwise fail.

Reply:
Well, I would have expected we parsed `perf record` error and error code directly?

```
% perf record -e blah echo 1
event syntax error: 'blah'
                     \___ parser error
Run 'perf list' for a list of valid events

 Usage: perf record [<options>] [<command>]
    or: perf record [<options>] -- <command> [<options>]

    -e, --event <event>   event selector. use 'perf list' to list available events

% echo $?
129

% perf record -e cycles:p echo 1 
Error:
cycles:p: PMU Hardware doesn't support sampling/overflow-interrupts. Try 'perf stat'

% echo $?                       
255

% perf record -e cycles echo 1       
1
[ perf record: Woken up 1 times to write data ]
[ perf record: Captured and wrote 0.013 MB perf.data (16 samples) ]

% echo $?                       
0
```

Yes, I understand it actually does not test whether the counters _report_ non-zero values, but that's okay. We only do a light-weight validation that `perf record` even accepts these.

Reply:
Failing hard when some events cannot be supported is also proper, I'd think. So there is no need to verify them one by one. Keep it as simple as possible?

Reply:
The thing is that `perf record` does not always fail when it doesn't work. For example in container ([example](https://github.com/galderz/github-actions/actions/runs/10723692355/job/29737549503)):

```
$ perf record --event cycles --output perf-record-validate.data echo 1
WARNING: Kernel address maps (/proc/{kallsyms,modules}) are restricted,
check /proc/sys/kernel/kptr_restrict and /proc/sys/kernel/perf_event_paranoid.

Samples in kernel functions may not be resolved if a suitable vmlinux
file is not found in the buildid cache or in the vmlinux path.

Samples in kernel modules won't be resolved at all.

If some relocation was applied (e.g. kexec) symbols may be misresolved
even with a suitable vmlinux or kallsyms file.

Couldn't record kernel reference relocation symbol
Symbol resolution may be skewed if relocation was used (e.g. kexec).
Check /proc/kallsyms permission or run as root.
1
[ perf record: Woken up 1 times to write data ]
[ perf record: Captured and wrote 0.001 MB perf-record-validate.data (3 samples) ]

$ echo $?
0

$ perf report -q -i perf-record-validate.data --stdio | grep cycles | wc -l
0
```

A lot of the issues I was having was because I assumed that in container `perf record` would fail but it doesn't, and later on you get errors saying that no events were recorded.

And in any case, if `perf record` fails, that's already handled by the current code when it calls `!failMsg.isEmpty()`.

> Failing hard when some events cannot be supported is also proper, I'd think. So there is no need to verify them one by one. Keep it as simple as possible?

I'll try to craft a command that checks all the events in one go and if any is missing fail hard.

Reply:
Oh, here's an interesting thing on containers. Some are outright not supported and `perf record` command fails:

```
$ perf record --event cycles,branches --output perf-record-validate.data echo 1
Error:
The branches event is not supported.
```

Reply:
`cpu-clock` has the same issue as `cycles`. `perf list` in container shows as supported and perf record works exits with 0:

```
$ perf record --event cycles,cpu-clock --output perf-record-validate.data echo 1

$ echo $?
0
```

But then no events:

```
$ perf report -q -i perf-record-validate.data --stdio | grep -E 'cycles|cpu-clock' | wc -l
0
```

So I'm thinking of check that the `perf report ... | wc -l` returns the count of requested event names, otherwise fail. That way is done with a single command.